monero/wallet_client.go

```diff
@@ -372,17 +550,48 @@ func (c *walletClient) Endpoint() string {
 	return c.endpoint
 }
 
+// Close kills the monero-wallet-rpc process closing the wallet. It is designed to only be
+// called a single time from a single go process.
 func (c *walletClient) Close() {
-	c.Lock()
-	defer c.Unlock()
-	if c.rpcProcess != nil {
-		p := c.rpcProcess
-		c.rpcProcess = nil
-		err := p.Kill()
-		if err == nil {
-			_, _ = p.Wait()
+	if c.rpcProcess == nil {
+		return // no monero-wallet-rpc instance was created
+	}
+	p := c.rpcProcess
+	err := c.wRPC.StopWallet()
+	if err != nil {
+		log.Warnf("StopWallet errored: %s", err)
+		err = p.Kill() // uses, SIG-TERM, which monero-wallet-rpc has a handler for
+	}
+	// If err is nil at this point, the process existed, and we block until the child
+	// process exits. (Note: kill does not error when signaling an exited, but non-reaped
+	// child.)
+	if err == nil {
+		_, _ = p.Wait()
+	}
+
```

Comment:
add a `c.rpcProcess == nil` in case this function is ever called twice? is there ever a chance of that happening?

Reply:
The code in this PR will never assign `rpcProcess` nil. The only way it can be nil, is if it was never assigned from the start (handled at the beginning of the function). There is nowhere that should be calling `Close()` more than once, but if it happens by accident, the 2nd `StopWallet()` will fail causing the `p.Kill()` line above to execute. Then the `Kill()` would also fail, as the process ID has already been reaped by `Wait()`, so we won't call `Wait()` a second time. What I describe above is the scenario where one `Close()` completed before the next one started. Even if both `Close()` calls were executed simultaneously, the above code should be safe from any adverse behavior.